kic_module.c

```diff
@@ -0,0 +1,198 @@
+{
+	pr_info("KIC: Device Opened.\n");
+	try_module_get(THIS_MODULE);
+	return 0;
+}
+/*
+ * maintains reference count, invoked by ioctl protocol
+ */
+static int kic_device_release(struct inode *inode, struct file *file)
+{
+	pr_info("KIC: Device Freed.\n");
+	module_put(THIS_MODULE);
+	return 0;
+}
+/*
+ * maintains exclusive access to the device and provides communication with
+ * userspace
+ */
+static long kic_device_ioctl(struct file *file, unsigned int ioctl_mode,
+			     unsigned long ioctl_carry)
+{
+	if (atomic_cmpxchg(&already_open, CDEV_NOT_USED, CDEV_EXCLUSIVE_OPEN)) {
+		pr_alert("KIC: Device is busy.\n");
+		return -EBUSY;
+	}
+
+	pr_info("KIC: Activating IOCTL protocol.\n");
+	switch (ioctl_mode) {
+	case KIC_RESET_QUERY: {		// option for resetting counter
+		kic_reset();
+		pr_info("KIC: Counter has been reset.\n");
+		break;
+	}// case closed
+	case KIC_GET_COUNT_QUERY: {	// option for retrieving count
+		long long kic_carry = atomic64_read(&kic_interrupt_count);
+
+		if (copy_to_user((void __user *)ioctl_carry, (void *)&kic_carry,
+				 sizeof(long long)) != 0)
+				pr_alert("KIC: Copy to user failed.\n");
+			else
+				pr_info("KIC: Count retrieved at %lld.\n",
+					 kic_carry);
+		break;
+	}// case closed
+	case KIC_GET_TIME_QUERY: {	// option for retrieving time and date
+		if (kic_time == 0) {
+			pr_alert("KIC: Timer has not yet been reset.\n");
+			put_user(-1, (long int __user *)ioctl_carry);
+			break;
+		}
+		put_user(kic_time, (long int __user *)ioctl_carry);
+		pr_info("KIC: Time retrieved.\n");
+		break;
+	}// case closed
+	}// switch closed
+	atomic_set(&already_open, CDEV_NOT_USED);
+	return 0;
+}
+
+static struct file_operations fops = {
+	.owner = THIS_MODULE,
+	.open = kic_device_open,
+	.release = kic_device_release,
+	.unlocked_ioctl = kic_device_ioctl,
+};
+/*
+ * purpose specific functions
+ *
+ * increases count upon detecting a valid interrupt
+ */
+static irqreturn_t kic_increment(int irq, void *dev_id)
+{
+	atomic64_inc(&kic_interrupt_count);
+	return IRQ_HANDLED;
+}
+
+/*
+ * resets counter and saves time data when invoked
+ */
+static void kic_reset(void)
+{
+	kic_time = ktime_get_real();
+	atomic64_set(&kic_interrupt_count, 0);
+}
+
+/*
+ * init and exit
+ */
+static int __init kic_init(void)
+{
+	int irq_ret;
+
+	irq_ret = register_chrdev(KIC_MAJOR, KIC_DEVICE_NAME, &fops);
+	if (irq_ret < 0) {
+		pr_alert("KIC: Registering device failed with error code: %d\n",
+			 irq_ret);
+		goto return_status;
+	}
+	pr_info("KIC: Device assigned major number %d\n", KIC_MAJOR);
+
+	kic_cls = class_create(THIS_MODULE, KIC_DEVICE_NAME);
+	if ((void *)kic_cls == ERR_PTR) {
+		pr_alert("KIC: Class creation failed.\n");
+		goto class_fail;
+	}
+	/*
+	 * may be slightly confusing, casting required for comparison,
+	 * pointer returned from device_create is useful only for error handling
+	 */
+	if ((void *)(device_create(kic_cls, NULL, MKDEV(KIC_MAJOR, 0), NULL,
```

Comment:
Whatever MKDEV(KIC_MAJOR, 0) does you use it in 3 places, maybe think how to not repeat it.

Reply:
will be obfuscated with a macro in next commit